File: app/src/test/java/com/ruideraj/secretelephant/main/MainViewModelTest.java
package com.ruideraj.secretelephant.main;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;

import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.ruideraj.secretelephant.AccountManager;
import com.ruideraj.secretelephant.R;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.ArgumentMatchers;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class MainViewModelTest {

    @Rule
    public InstantTaskExecutorRule instantExecutorRule = new InstantTaskExecutorRule();

    @Mock
    private AccountManager mAccountManager;

    @Mock
    private GoogleSignInAccount mAccount;

    private MainViewModel mMainViewModel;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);

        mMainViewModel = new MainViewModel(mAccountManager);
    }

    @Test
    public void startWithAccount() {
        when(mAccountManager.getAccount()).thenReturn(mAccount);

        mMainViewModel.start();

        assertNotNull(mMainViewModel.signedIn.getValue());
        assertTrue(mMainViewModel.signedIn.getValue());
    }

    @Test
    public void startNoAccount() {
        when(mAccountManager.getAccount()).thenReturn(null);

        mMainViewModel.start();

        assertNotNull(mMainViewModel.signedIn.getValue());
        assertFalse(mMainViewModel.signedIn.getValue());
    }

    @Test
    public void signOut() {
        mMainViewModel.signOut();
        verify(mAccountManager, times(1))
                .signOut(ArgumentMatchers.any(AccountManager.AccountListener.class));
    }

    @Test
    public void signOutSuccess() {
        mMainViewModel.onSignOutSuccess();

        assertNotNull(mMainViewModel.signedIn.getValue());
        assertFalse(mMainViewModel.signedIn.getValue());

        assertNotNull(mMainViewModel.signOutMessage.getValue());
        assertEquals(R.string.main_menu_signed_out, (int) mMainViewModel.signOutMessage.getValue());
    }

    @Test
    public void signOutFailure() {
        mMainViewModel.onSignOutFailure();

        assertNotNull(mMainViewModel.signOutMessage.getValue());
        assertEquals(R.string.main_menu_sign_out_failed, (int) mMainViewModel.signOutMessage.getValue());
    }
}

File: app/src/test/java/com/ruideraj/secretelephant/main/MainViewModelTest.kt
package com.ruideraj.secretelephant.main

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.ruideraj.secretelephant.AccountManager
import com.ruideraj.secretelephant.R
import io.mockk.*
import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test

class MainViewModelTest {

    @get:Rule
    val instantExecutorRule = InstantTaskExecutorRule()

    private var accountManager: AccountManager = mockk()

    private var account: GoogleSignInAccount = mockk()

    private lateinit var viewModel: MainViewModel

    @Before
    fun setup() {
        viewModel = MainViewModel(accountManager)
    }

    @Test
    fun mainViewModel_start_signedInWhenAccountPresent() {
        every { accountManager.getAccount() } returns account

        viewModel.start()

        assertThat(viewModel.signedIn.value, equalTo(true))
    }

    @Test
    fun mainViewModel_start_signedOutWhenAccountNotPresent() {
        every { accountManager.getAccount() } returns null

        viewModel.start()
        assertThat(viewModel.signedIn.value, equalTo(false))
    }

    @Test
    fun mainViewModel_signOut_signOutSucceeds_signedOutWithMessage() {
        val listenerSlot = slot<AccountManager.AccountListener>()
        every { accountManager.signOut(capture(listenerSlot)) } just Runs

        viewModel.signOut()

        verify { accountManager.signOut(any()) }

        listenerSlot.captured.onSignOutSuccess()

        assertThat(viewModel.signedIn.value, equalTo(false))
        assertThat(viewModel.signOutMessage.value, equalTo(R.string.main_menu_signed_out))
    }
    
    @Test
    fun mainViewModel_signOut_signOutFails_showsSignOutFailMessage() {
        val listenerSlot = slot<AccountManager.AccountListener>()
        every { accountManager.signOut(capture(listenerSlot)) } just Runs

        viewModel.signOut()

        verify { accountManager.signOut(any()) }

        listenerSlot.captured.onSignOutFailure()

        assertThat(viewModel.signOutMessage.value, equalTo(R.string.main_menu_sign_out_failed))
    }
}